File: als1/hashtable/hashtable.py
class HashTable:
    def __init__(self, sz, stp):
        self.size = sz
        self.step = stp
        self.slots = [None] * self.size

    def hash_fun(self, value):
        value = value.encode("utf-8")
        sum_bytes = 0
        for byte in value:
            sum_bytes += byte
        return sum_bytes % self.size

    def seek_slot(self, value):
        if not self.slots.count(None):
            return None
        index = self.hash_fun(value)
        if not self.slots[index]:
            return index
        new_index = (index + self.step) % self.size
        while new_index != index:
            if not self.slots[new_index]:
                return new_index
            new_index = (new_index + self.step) % self.size
        return None

    def put(self, value):
        index = self.seek_slot(value)
        if index is not None:
            self.slots[index] = value
        return index

    def find(self, value):
        index = self.hash_fun(value)
        if self.slots[index] == value:
            return index
        new_index = (index + self.step) % self.size
        while new_index != index:
            if self.slots[new_index] == value:
                return new_index
            new_index = (new_index + self.step) % self.size
        return None

File: asd_go/hashtable/hashtable.go
package main

import (
	"os"
	"strconv"
)

type HashTable struct {
	size   int
	step   int
	count  int
	slots  []string
	filled []bool
}

func Init(sz int, stp int) HashTable {
	ht := HashTable{size: sz, step: stp, slots: nil}
	ht.slots = make([]string, sz)
	ht.filled = make([]bool, sz)
	return ht
}

func (ht *HashTable) HashFun(value string) int {
	hash := 1
	for _, c := range value {
		hash += int(c)
	}
	return hash % ht.size
}

func (ht *HashTable) SeekSlot(value string) int {
	if ht.full() {
		return -1
	}
	key := ht.seek(value)
	return key
}

func (ht *HashTable) full() bool {
	return ht.count == ht.size
}

func (ht *HashTable) Put(value string) int {
	key := ht.SeekSlot(value)
	if key != -1 && !ht.filled[key] {
		ht.filled[key] = true
		ht.slots[key] = value
		ht.count++
	}
	return key
}

func (ht *HashTable) Find(value string) int {
	key := ht.seek(value)

	if key == -1 {
		return -1
	}

	if ht.slots[key] != value {
		return -1
	}
	return key
}

func (ht *HashTable) seek(value string) int {
	key := ht.HashFun(value)
	start := key
	fromStart := false
	for ht.filled[key] {
		if ht.slots[key] == value {
			return key
		}
		key = (ht.step + key) % ht.size

		if key <= start {
			fromStart = true
		}

		if fromStart && key >= start {
			return -1
		}
	}
	return key
}